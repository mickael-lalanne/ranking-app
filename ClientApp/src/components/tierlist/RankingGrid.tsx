import React from 'react';
import { css } from '@emotion/css';
import { TIERS_COLORS, Template } from '../../models/Template';
import { RankedElement, Element } from '../../models/Element';
import ElementPreview from '../shared/ElementPreview';

const RankingGrid = ({ template, dropHandler, rankedElements }: {
    template?: Template,
    rankedElements: RankedElement[],
    dropHandler: (tierId: number, rank: number) => void
}) => {
    /**
     * @param {number} tierId necessary for the drop handler
     * @returns {React.JSX.Element[]} the 5 cells contained in a tier
     */
    const TierCells = (tierId: number): React.JSX.Element[] => {
        const cells: React.JSX.Element[] = [];

        /**
         * Check if an element is ranked in the cell
         * @param {number} position the cell position 
         * @returns {React.JSX.Element} an element if the cell contains one
         */
        const ElementInCell = (position: number): React.JSX.Element => {
            const elementInCell: RankedElement | undefined = rankedElements.find(
                rankedElt => rankedElt.tierId === tierId && rankedElt.position === position
            );
            
            const elementObject: Element | undefined = template?.elements.find(
                elt => elt.id === elementInCell?.elementId
            );

            if (elementInCell && elementObject) {
                return (
                    <ElementPreview element={elementObject} />
                );
            }
            return <></>;
        };

        for (let cell = 0; cell < 5; cell++) {
            cells.push(
                <div
                    className={tier_cell_style}
                    key={cell}
                    onDrop={() => dropHandler(tierId, cell)}
                    onDragOver={e => e.preventDefault()}
                    onDragEnter={e => e.preventDefault()}
                >
                    {ElementInCell(cell)}
                </div>
            );
        }

        return cells;
    }

    /**
     * @returns {React.JSX.Element[]} all tiers that the template has
     */
    const TiersLines = (): React.JSX.Element[] => {
        if (!template) { return []; }

        const lines: React.JSX.Element[] = [];

        template.tiers.forEach(tier => {
            lines.push(
                <div className={tier_line_style} style={{ backgroundColor: TIERS_COLORS[tier.rank]}} key={tier.id}>
                    {TierCells(tier.id!)}
                </div>
            );
        });

        return lines;
    };

    return(<>
        <div className={tiers_grid_style}>
            {TiersLines()}
        </div>
    </>);
};

export default RankingGrid;

/**
 * CSS STYLES
 */
const tiers_grid_style = css({
    display: 'flex',
    flexDirection: 'column',
    maxWidth: '530px',
    margin: 'auto'
});

const tier_line_style = css({
    display: 'flex',
    flexDirection: 'row',
    width: '100%'
});

const tier_cell_style = css({
    width: '20%',
    aspectRatio: '1 / 1'
});
