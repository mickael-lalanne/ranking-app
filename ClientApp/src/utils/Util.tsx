import React from "react";

// All items that are not yet saved in the database starts with the following identifier
const TMP_ID_IDENTIFIER = '99999';

/**
 * Generate a temporary for items that are not yet saved in database
 * @returns {string} a random temporary id
 */
export const generateRandomId = (): string => {
    const randomNumber: number = Math.floor(Math.random() * 1000000000);
    return TMP_ID_IDENTIFIER + randomNumber;
};

/**
 * Check if an id is temporary
 * @param {string} id the id to check
 * @returns {boolean} true if the id is temporary
 */
export const isTemporaryId = (id: string): boolean => {
    return id.startsWith(TMP_ID_IDENTIFIER);
}

export interface ResizedImage {
    source: string; // the data URL of the image
    name: string;
}

/**
 * Resize an image while keeping its ratio
 * This function is called in the "onChange" event of input file
 * So we use the FileReader api to get the base image dimensions
 * @param {File} imageFile the image file to resize
 * @returns {Promise<ResizedImage>} an object containing the resized image as data URL and its name
 */
export const resizeImage = (imageFile: File): Promise<ResizedImage> => {
    const MAX_SIZE: number = 150;
    const reader = new FileReader();

    return new Promise ((resolve) => {
        reader.onload = () => {
            const tmpImg: HTMLImageElement = new Image();
    
            tmpImg.onload = () => {
                const resizeRatio = Math.min(MAX_SIZE / tmpImg.width, MAX_SIZE / tmpImg.height);

                // Initialize the canvas
                const canvas = document.createElement("canvas");
                const ctx = canvas.getContext('2d');
    
                // Set width and height
                canvas.width = tmpImg.width * resizeRatio;
                canvas.height = tmpImg.height * resizeRatio;
    
                // Draw image and export to a data-uri
                ctx!.drawImage(tmpImg, 0, 0, canvas.width, canvas.height);
    
                const dataURI = canvas.toDataURL();
                resolve({ source: dataURI, name: imageFile.name });
            };

            tmpImg.src = URL.createObjectURL(imageFile);
        }
        reader.readAsDataURL(imageFile);
    });

};

/**
 * @param array an array containing a createdAt field
 * @returns {Partial<{ createdAt: string }>[]} the same array but sorted by creation date
 */
export const sortByCreationDate = (array: Partial<{ createdAt: string }>[]) => {
    // Use slice() to avoid reference errors
    return array.slice().sort((a, b) => 
        new Date(a.createdAt!).getTime() - new Date(b.createdAt!).getTime()
    );
};

/**
 * Called in the template and tierlist editors
 * Returns a value for the MUI Tooltip component
 * @param {string?} content the text that should be displayed in the tooltip 
 * @returns {'' | React.JSX.Element | undefined} a span including the text if defined, otherwise undefined
 */
export const getTooltipTitleForSaveButtons = (
    content?: string
): '' | React.JSX.Element | undefined => {
    return content && <span style={{ fontSize: '15px'}}>{content}</span>;
}
