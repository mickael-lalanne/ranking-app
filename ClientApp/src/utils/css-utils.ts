export const ELEMENT_SIZE: string = '75px';
