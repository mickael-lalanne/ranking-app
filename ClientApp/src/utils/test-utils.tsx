import React, { PropsWithChildren } from 'react';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { setupStore } from '../app/store';
import type { AppStore, RootState } from '../app/store';
import { ThemeOptions, ThemeProvider, createTheme } from '@mui/material/styles';
import '@testing-library/jest-dom';
import { DndProvider } from 'react-dnd';
import { TouchBackend } from 'react-dnd-touch-backend';

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
export interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
    preloadedState?: Partial<RootState>;
    store?: AppStore;
}

// Mock MUI theme
const RANKING_APP_THEME_MOCK: ThemeOptions = {
    defaultRankingTheme: {
        primary: '',
        primaryLight: '',
        light: '',
        dark: '',
        info: '',
    },
};
const MockTheme = ({ children }: any) => {
    const theme = createTheme(RANKING_APP_THEME_MOCK);
    return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        // Automatically create a store instance if no store was passed in
        store = setupStore(preloadedState),
        ...renderOptions
    }: ExtendedRenderOptions = {}
) {
    function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
        return (
            <MockTheme>
                <Provider store={store}>
                    <DndProvider
                        backend={TouchBackend}
                        options={{ enableMouseEvents: true }}
                    >
                        {children}
                    </DndProvider>
                </Provider>
            </MockTheme>
        );
    }

    // Return an object with the store and all of RTL's query functions
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
