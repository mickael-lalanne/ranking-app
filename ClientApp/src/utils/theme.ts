import { Theme, createTheme, alpha, ThemeOptions } from "@mui/material";

interface RankingTheme {
    primary: string;
    primaryLight: string;
    light: string;
    dark: string;
    info: string;
};
const PRIMARY_COLOR = '#BB2525';

declare module '@mui/material/styles' {
    interface Theme {
        defaultRankingTheme: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
    // allow configuration using `createTheme`
    interface ThemeOptions {
        defaultRankingTheme?: {
            primary: string;
            primaryLight: string;
            light: string;
            dark: string;
            info: string;
        };
    }
}

export const THEMES_COLORS: ThemeOptions['defaultRankingTheme'] = {
    primary: PRIMARY_COLOR,
    primaryLight: '#FF6969',
    light: '#FFF5E0',
    dark: '#141E46',
    info: '#9ED9EB'
}

export const RANKING_APP_THEME: Theme = createTheme({
    defaultRankingTheme: THEMES_COLORS,
    palette: {
        primary: {
            main: alpha(PRIMARY_COLOR, 0.7),
            light: alpha(PRIMARY_COLOR, 0.5),
            dark: alpha(PRIMARY_COLOR, 0.9),
        }
    }
});
