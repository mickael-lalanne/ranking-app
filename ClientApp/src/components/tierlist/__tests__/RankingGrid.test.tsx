import * as React from 'react';
import { renderWithProviders } from '../../../utils/test-utils';
import RankingGrid, { RankingGridProps } from '../RankingGrid';

const rankedEltId1: string = '0';
const rankedEltId2: string = '1';

const DEFAULT_PROPS: RankingGridProps = {
    rankedElements: [
        { tierId: '0', elementId: rankedEltId1, position: 0 },
        { tierId: '2', elementId: rankedEltId2, position: 4 },
    ],
    template: {
        id: 'my_template_id',
        name: 'template name',
        tiers: [
            { id: '0', name: 'tier 0', rank: 0 },
            { id: '1', name: 'tier 1', rank: 1 },
            { id: '2', name: 'tier 2', rank: 2 },
        ],
        elements: [
            { id: rankedEltId1, name: 'ranked-element-1', image: 'ranked-img' },
            { id: rankedEltId2, name: 'ranked-element-2', image: 'ranked-img' },
            { id: 'x', name: 'not-ranked-element', image: 'unranked-img' },
        ],
    },
    readonly: false,
};

const RankingGridTest = (props: RankingGridProps) => {
    return (
        <RankingGrid
            rankedElements={props.rankedElements}
            template={props.template}
            dropHandler={props.dropHandler}
            readonly={props.readonly}
        />
    );
};

describe('Screenshot', () => {
    it('takes default screenshot', () => {
        const component = renderWithProviders(
            RankingGridTest(DEFAULT_PROPS)
        );

        expect(component.baseElement).toMatchSnapshot();
    });
});

// FYI: d&d behaviors are tested in cypress with e2e tests
