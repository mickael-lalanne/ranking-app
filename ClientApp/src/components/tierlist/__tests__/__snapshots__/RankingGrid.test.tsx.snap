// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Screenshot takes default screenshot 1`] = `
<body>
  <div>
    <div
      class="css-10cvn1p"
      data-cy="ranking-grid"
      data-testid="ranking-grid"
    >
      <div
        class="css-1ik4laa"
        data-cy="ranking-grid-tier-line"
        style="background-color: rgb(255, 140, 127);"
      >
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        >
          <div
            class="css-13ambje"
            data-cy="element-preview"
            data-testid="element-preview"
            draggable="true"
            style="opacity: 1;"
          >
            <img
              class="css-40zi5q"
              data-cy="element-preview-img"
              src="https://res.cloudinary.com/ranking-app/image/upload/ranked-img"
            />
          </div>
        </div>
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
      </div>
      <div
        class="css-1ik4laa"
        data-cy="ranking-grid-tier-line"
        style="background-color: rgb(255, 192, 125);"
      >
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
      </div>
      <div
        class="css-1ik4laa"
        data-cy="ranking-grid-tier-line"
        style="background-color: rgb(254, 247, 130);"
      >
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        />
        <div
          class="css-1u0sc0w"
          data-cy="element-in-cell"
        >
          <div
            class="css-13ambje"
            data-cy="element-preview"
            data-testid="element-preview"
            draggable="true"
            style="opacity: 1;"
          >
            <img
              class="css-40zi5q"
              data-cy="element-preview-img"
              src="https://res.cloudinary.com/ranking-app/image/upload/ranked-img"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
